File: java/medium/152_maximum_product_subarray.java
// https://leetcode.com/problems/maximum-product-subarray/

package java.medium;

class MaximumProductSubarray {
    public int maxProduct(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }

        int result = nums[0];
        int min = 1;
        int max = 1;

        for (int n : nums) {
            int tmp = max * n;
            max = Math.max(n, Math.max(min * n, tmp));
            min = Math.min(n, Math.min(min * n, tmp));

            result = Math.max(result, max);
        }

        return result;
    }
}

File: go/medium/152_maximum_product_subarray.go
package medium

func MaxProductSubarray(nums []int) int {
	n := len(nums)
	if n == 0 {
		return 0
	}

	currMax := nums[0]
	currMin := nums[0]
	result := currMax

	for i := 1; i < n; i++ {
		curr := nums[i]
		tmpMax := max(curr, max(currMax*curr, currMin*curr))
		currMin = min(curr, min(currMax*curr, currMin*curr))
		currMax = tmpMax

		result = max(result, currMax)
	}

	return result
}
